presto-main/src/main/java/com/facebook/presto/cost/ComposableStatsCalculator.java

```diff
@@ -66,21 +66,26 @@ public ComposableStatsCalculator(List<Rule> rules)
     @Override
     public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, Map<Symbol, Type> types)
     {
-        Iterator<Rule> ruleIterator = getCandidates(node).iterator();
+        Iterator<Rule<?>> ruleIterator = getCandidates(node).iterator();
         while (ruleIterator.hasNext()) {
-            Rule rule = ruleIterator.next();
-            Optional<PlanNodeStatsEstimate> calculatedStats = rule.calculate(node, sourceStats, lookup, session, types);
+            Rule<?> rule = ruleIterator.next();
+            Optional<PlanNodeStatsEstimate> calculatedStats = calculatedStats(rule, node, sourceStats, lookup, session, types);
```

Comment:
Why not just 
```
@SuppressWarnings({"rawtypes", "unchecked"})
Optional<PlanNodeStatsEstimate> calculatedStats = ((Rule) rule).calculate(node, sourceStats, lookup, session, types);
```
?

Reply:
No, using raw types is a code smell and is typically a sign that the API or types are not appropriate.

From the Java Language Specification:

> The use of raw types is allowed only as a concession to compatibility of legacy code. The use of raw types in code written after the introduction of generics into the Java programming language is strongly discouraged. It is possible that future versions of the Java programming language will disallow the use of raw types.

Reply:
Well, unchecked is also a code smell, but i understand suppressing one is better than suppressing two.

Reply:
@martint @findepi Piotr, I was following the code in 

https://github.com/mbasmanova/presto/blob/955e46339267780e9176e1289954439213fefe3f/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java#L141

I think it should be fine.